### PR TITLE
🌿 introduce Fern Definition + SDKs

### PR DESCRIPTION
**Pull request by dsinghvi**
### Dependencies

There are no dependencies for this PR. 

### Description

In this PR, we introduce the fern definition for Metriport's API directly into the monorepo. Here are some relevant commands: 
- `fern check` will check the API definition and ensure its valid. If the definition is valid, Fern will be able to generate SDKs!
- `fern generate --group local` will generate all the source code for the SDKs inside of the repo. 
- `fern generate --group publish --version <version>` will actually publish the SDKs to their package managers. 

**Note** I've added a Github action that will check that the generated source code for the SDKs is always up to date. This action will run on all pull requests. 

### Release Plan

Before we merge this PR, we need to add a couple secrets to this repo. Specifically, `MAVEN_USERNAME`, `MAVEN_PASSWORD`, `NPM_TOKEN` and `PYPI_TOKEN`. Adding these secrets will allow you to run `fern generate --group publish`. 

Once this PR is merged, we can delete the `metriport-api` repo that houses the Fern definition today.

### CODE REVIEW

**Review by leite08 — COMMENTED** [review]
I did a first pass, haven't finished reviewing all endpoints (focused mostly on the general approach and the `organization.yml`) - looking forward to review the rest once we go through these:
- When we introduce the Express generator we'll have access to the types defined in the Fern definition there, on the "Services", right?
- Also, w/ Express generator, will the error messages come from the Fern definition? I don't see them stored there.. or will it use what's on the `docs` property?
- In the event we need to make changes to the SDK but want to keep backwards compatibility, for example renaming an endpoint's name from `list` to `search` or `getAll`, is it possible to have more than one name for the same endpoint? Do we need to create two? If the latter, what about Express generator, is there a way to not actually duplicate the endpoint?
- How to go about custom code on the SDK? Like, to upload files to S3 we're using pre-signed URLs, but that returns a URL and form fields to be used as FormData - we would like to be able to abstract this from our users at some point if possible; this would require some logic on the SDK.
- The TS SDK's `package.json` will get generated on the first publication? It would be good to review it before it gets published.
- With this PR merged we can plan the migration of the existing SDK into the one generated here, right?
   - And so, with this PR merged it also means that any changes on the contract of the endpoints starting with `/connect`, `/users`, and `/medical` need to be replicated to the Fern definition, right?
- Product (just so I can keep track while reviewing, we can move to another place you prefer)
   - Is there a Yaml schema we should point our VSCode to to get schema/compilation validation while we work on the schemas? I tried this on `settings.json` but it didn't work:
      ```
      "yaml.schemas": {
         "https://github.com/fern-api/fern/blob/main/fern.schema.json": "fern/**/*.yml"
      }
      ```
   - can we cmd/ctrl + click on Fern definitions and go to the referenced file (like `../commons.yml` - can do that w/ GH's workflows that are `.yml` files as well, is this an extension thing?)
   - when I run `fern generate --local --keepDocker`
      - it always creates a new container - I thought that `--keepDocker` would keep the docker container around
      - it really doesn't work without internet, even though the `--local` is there

**Comment by dsinghvi** [comment]
@leite08 thanks for the in-depth review! 

> When we introduce the Express generator we'll have access to the types defined in the Fern definition there, on the "Services", right?

This is correct. I haven't added the Express generator yet since its part of a different stage. 

> Also, w/ Express generator, will the error messages come from the Fern definition? I don't see them stored there.. or will it use what's on the docs property?

This depends on what you want, you can use the `literal` type in Fern and add the correct error message there. In this case, the error messages will come from the fern definition and they will show up in the docs as well. The SDK behavior won't really be affected. 


> In the event we need to make changes to the SDK but want to keep backwards compatibility, for example renaming an endpoint's name from list to search or getAll, is it possible to have more than one name for the same endpoint? Do we need to create two? If the latter, what about Express generator, is there a way to not actually duplicate the endpoint?

Great question, most of our customers use the release of the Fern SDK to force their users to upgrade. That way they can get onto the "latest and greatest" and you can remove any cruft or tech debt from the SDK. So, if possible, I'd be curious what you thought of naming the methods appropriately and not being backwards compatible. 

If backwards compatibility was important, I think the best thing here would be for us to build an additional feature where you could specify `alternative-sdk-method-names` in the spec. 

> The TS SDK's package.json will get generated on the first publication? It would be good to review it before it gets published.

The generator would only generate it when in `publish` mode but I can track a change for this. Would you be open to doing this in a follow up? 

> With this PR merged we can plan the migration of the existing SDK into the one generated here, right?
> And so, with this PR merged it also means that any changes on the contract of the endpoints starting with /connect, /users, and /medical need to be replicated to the Fern definition, right?

That's right -- I would love to do that in a follow up PR as its riskier and we can isolate the changes/go back and forth.

**Comment by dsinghvi** [comment]
> Is there a Yaml schema we should point our VSCode to to get schema/compilation validation while we work on the schemas? I tried this on settings.json but it didn't work:

Did you have the vscode yaml plugin installed? One thing that works well is to have that plugin installed and add the following comment at the top of the file 

```yaml
# yaml-language-server: $schema=https://raw.githubusercontent.com/fern-api/fern/main/fern.schema.json

...definition
```

> can we cmd/ctrl + click on Fern definitions and go to the referenced file (like ../commons.yml - can do that w/ GH's workflows that are .yml files as well, is this an extension thing?)

You can't yet, but this is a very reasonable request. Usually you need to get a language server working for this, which we haven't built yet. 

> when I run fern generate --local --keepDockerit always creates a new container
>  - I thought that --keepDocker would keep the docker container around
>  - it really doesn't work without internet, even though the --local is there

Yes it runs a new docker. `--keepDocker` keeps the stopped docker image around for debugging, otherwise it'll delete it by default. It should work without internet as long as the docker is on your system (if not it wont).

File: fern/definition/medical/organization.yml
Comment (dsinghvi) [reply fern/definition/medical/organization.yml:44]: Great -- I defer to you on the naming here (you know better than me). I can make the changes but it might be faster if you went an updated, thoughts?

File: fern/definition/medical/organization.yml
Comment (dsinghvi) [reply fern/definition/medical/organization.yml:27]: Makes sense.

File: fern/definition/medical/organization.yml
Comment (dsinghvi) [reply fern/definition/medical/organization.yml:77]: Sounds good

File: fern/definition/commons.yml
Comment (dsinghvi) [reply fern/definition/commons.yml:19]: That is currently not supported, but something we intend to add. This hasn't been a big issue from the SDK side because folks usually add validation on their end in the server.

File: fern/definition/commons.yml
Comment (dsinghvi) [reply fern/definition/commons.yml:13]: See https://github.com/metriport/metriport/pull/1029#discussion_r1373830474

File: fern/definition/commons.yml
Comment (dsinghvi) [reply fern/definition/commons.yml:7]: Not supported (see https://github.com/metriport/metriport/pull/1029#discussion_r1373830474)

File: packages/sdks/java/core/ClientOptions.java
Comment (dsinghvi) [reply packages/sdks/java/core/ClientOptions.java:27]: Thats a bug, I should remove those headers

**Comment by dsinghvi** [comment]
**Follow-up** Specify organization, license, github repo so that it gets correctly populated in `package.json`, `pyproject.toml`, etc.

**Comment by dsinghvi** [comment]
Add comment to all fern definition files: 

# yaml-language-server: $schema=https://raw.githubusercontent.com/fern-api/fern/main/fern.schema.json

**Comment by dsinghvi** [comment]
Should fix: https://github.com/fern-api/fern/issues/2003

**Comment by leite08** [comment]
Secrets available to this repo:
- [x] `MAVEN_USERNAME`
- [x] `MAVEN_PASSWORD`
- [x] `NPM_TOKEN`
- [ ] `PYPI_TOKEN`

File: fern/definition/api.yml
Comment (leite08) [inline fern/definition/api.yml:8]: How to deal with the fact there are diff base URLs for Medical and Devices API? One api spec per api (MAPI/DAPI)?

File: fern/definition/commons.yml
Comment (leite08) [reply fern/definition/commons.yml:19]: Link to the issue/ticket so we can keep track of it, please?

File: fern/definition/commons.yml
Comment (leite08) [reply fern/definition/commons.yml:7]: add function to validate this @ route level, maybe keep zod schemas around for now?

File: fern/definition/commons.yml
Comment (leite08) [reply fern/definition/commons.yml:13]: add function to validate this @ route level, maybe keep zod schemas around for now?

File: fern/definition/commons.yml
Comment (leite08) [reply fern/definition/commons.yml:19]: add function to validate this @ route level, maybe keep zod schemas around for now?

File: fern/definition/devices/__package__.yml
Comment (leite08) [inline fern/definition/devices/__package__.yml:1]: What does `__package__` mean? Is it like the representation of `/`, in this case `/devices`?

I'll leave the contents of this file to be reviewed after merged to our repo

File: fern/definition/devices/connect.yml
Comment (leite08) [inline fern/definition/devices/connect.yml:1]: ditto: review after merged

File: fern/definition/devices/users.yml
Comment (leite08) [inline fern/definition/devices/users.yml:1]: ditto: review after merged

File: fern/definition/medical/document.yml
Comment (leite08) [inline fern/definition/medical/document.yml:1]: ditto: review after merged

File: fern/definition/medical/facility.yml
Comment (leite08) [inline fern/definition/medical/facility.yml:1]: ditto: review after merged

File: fern/definition/medical/fhir.yml
Comment (leite08) [inline fern/definition/medical/fhir.yml:1]: ditto: review after merged

File: fern/definition/medical/patient.yml
Comment (leite08) [inline fern/definition/medical/patient.yml:1]: ditto: review after merged

**Review by leite08 — COMMENTED** [review]
Left a few comments/asks and a bunch to review after merged.

File: fern/definition/api.yml
Comment (dsinghvi) [reply fern/definition/api.yml:8]: We support different URLs per endpoint, i can make that change!

File: fern/definition/medical/facility.yml
Comment (dsinghvi) [reply fern/definition/medical/facility.yml:38]: Taken from docs: https://docs.metriport.com/medical-api/api-reference/facility/get-facility

```
The ID assigned to this Facility. This ID will be used to uniquely identify this Facility in medical documents.
```

File: fern/definition/medical/facility.yml
Comment (dsinghvi) [reply fern/definition/medical/facility.yml:111]: Taken from docs: https://docs.metriport.com/medical-api/api-reference/facility/get-facility

```
The ID assigned to this Facility. This ID will be used to uniquely identify this Facility in medical documents.
```

File: fern/definition/commons.yml
Comment (dsinghvi) [reply fern/definition/commons.yml:19]: Here's a link: https://github.com/fern-api/fern/issues/2246

File: .github/workflows/fern-check.yml
Comment (dsinghvi) [reply .github/workflows/fern-check.yml:6]: Yup -- added to `develop`. We are working on something with `alpha` releases but for now people do them by appending `-beta` or `-alpha` when they run fern generate and pass in the version.

**Review by leite08 — APPROVED** [review]

**Pull request merged**
